src/m3ninx/index/regexp.go

```diff
@@ -0,0 +1,323 @@
+			}
+			parsed.Sub[idx] = newRe
+			changed = changed || c
+		}
+	case syntax.OpQuest:
+		if len(parsed.Sub) > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[0] = newRe
+				changed = true
+			}
+		}
+	case syntax.OpStar:
+		if len(parsed.Sub) > 0 {
+			original := deepCopy(parsed)
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if !c {
+				return parsed, false, nil
+			}
+			newRegexp = &syntax.Regexp{
+				Op:    syntax.OpConcat,
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					&syntax.Regexp{
+						Op:    syntax.OpQuest,
+						Flags: parsed.Flags,
+						Sub: []*syntax.Regexp{
+							newRe,
+						},
+					},
+					original,
+				},
+			}
+			changed = true
+		}
+	case syntax.OpPlus:
+		if len(parsed.Sub) > 0 {
+			original := deepCopy(parsed)
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if !c {
+				return parsed, false, nil
+			}
+			return &syntax.Regexp{
+				Op:    syntax.OpConcat,
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					newRe,
+					&syntax.Regexp{
+						Op:    syntax.OpStar,
+						Flags: parsed.Flags,
+						Sub: []*syntax.Regexp{
+							original.Sub[0],
+						},
+					},
+				},
+			}, true, nil
+		}
+	case syntax.OpRepeat:
+		if len(parsed.Sub) > 0 && parsed.Min > 0 && parsed.Max > 0 {
+			original := deepCopy(parsed)
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if !c {
+				return parsed, false, nil
+			}
+			original.Min--
+			original.Max--
+			newRegexp = &syntax.Regexp{
+				Op:    syntax.OpConcat,
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					newRe,
+					original,
+				},
+			}
+			changed = true
+		}
+	}
+	return newRegexp, changed, nil
+}
+
+func deepCopy(ast *syntax.Regexp) *syntax.Regexp {
+	if ast == nil {
+		return nil
+	}
+	copied := *ast
+	copied.Sub = make([]*syntax.Regexp, 0, len(ast.Sub))
+	for _, r := range ast.Sub {
+		copied.Sub = append(copied.Sub, deepCopy(r))
+	}
+	if len(copied.Sub0) != 0 && copied.Sub0[0] != nil {
+		copied.Sub0[0] = deepCopy(copied.Sub0[0])
+	}
+	return &copied
+}
+
+var emptyStringRes = []*syntax.Regexp{
+	&syntax.Regexp{Op: syntax.OpEmptyMatch},
+	&syntax.Regexp{Op: syntax.OpQuest, Sub: []*syntax.Regexp{&syntax.Regexp{Op: syntax.OpEmptyMatch}}},
+	&syntax.Regexp{Op: syntax.OpPlus, Sub: []*syntax.Regexp{&syntax.Regexp{Op: syntax.OpEmptyMatch}}},
+	&syntax.Regexp{Op: syntax.OpStar, Sub: []*syntax.Regexp{&syntax.Regexp{Op: syntax.OpEmptyMatch}}},
+}
+
+func matchesEmptyString(ast *syntax.Regexp) bool {
+	if ast == nil {
+		return false
+	}
+	for _, empty := range emptyStringRes {
+		if ast.Equal(empty) {
```

Comment:
Hm I wonder if actually doing specific checks of the op type and then whether the subexpressions match OpEmptyMatch is a good idea?  i.e. if OpRepeat is also a similar case, then you don't want to compare the Min/Max fields of the regexp against your literal emptyStringRes OpRepeat regexp.

Reply:
yeah i re-wrote this recursively to effectively do the following:``` matchesEmptyString(ast): return true if {ast is OpEmptyMatch } or {ast is a unary op'd (?/*/+/repeat) of a ast that matchesEmptyString}```